File: exercise/684/solution.cpp
struct UnionFind {
    int root[1001];
    UnionFind() {
        for (int i = 0; i <= 1000; i++) root[i] = i;
    }
    int Find(int a) {
        if (root[a] != a)
            root[a] = Find(root[a]);
        return root[a];
    }
    void insert(int a, int b) {
        a = Find(a);
        b = Find(b);
        root[a] = b;
    }
    bool Same(int a, int b) {
        return Find(a) == Find(b);
    }
};

class Solution {
public:
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        UnionFind unionFind;
        for (int i = 0; i < edges.size(); i++) {
            if (unionFind.Same(edges[i][0], edges[i][1])) {
                return edges[i];
            } else {
                unionFind.insert(edges[i][0], edges[i][1]);
            }
        }
        return vector<int>();
    }
};

File: exercise/684/solution.py
class Union:

    def __init__(self, size):
        self.root = [i for i in range(size + 1)]

    def add(self, a, b):
        self.root[self.find(b)] = self.find(a)

    def find(self, a):
        if self.root[a] != a:
            self.root[a] = self.find(self.root[a])
        return self.root[a]

    def same(self, a, b):
        return self.find(a) == self.find(b)

class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        all_union = Union(len(edges))
        for edge in edges:
            if all_union.same(*edge):
                return edge
            else:
                all_union.add(*edge)
        return []
        
